client/src/main/java/org/eclipse/hono/client/impl/HonoConnectionImpl.java

```diff
@@ -492,14 +495,21 @@ private void reconnect(
             if (connectionFailureCause != null) {
                 log.debug("connection attempt failed", connectionFailureCause);
             }
-            final long reconnectInterval = Optional.ofNullable(clientOptions)
-                    .map(o -> o.getReconnectInterval())
-                    .orElse(Constants.DEFAULT_RECONNECT_INTERVAL_MILLIS);
-            log.trace("scheduling new connection attempt in {}ms ...", reconnectInterval);
-            // give Vert.x some time to clean up NetClient
-            vertx.setTimer(reconnectInterval, tid -> {
+            // apply exponential backoff with jitter
+            final long currentMaxDelay = (long) Math.pow(2, reconnectAttempt - 1)
+                    * clientConfigProperties.getReconnectDelayIncrement();
+            final long reconnectInterval = currentMaxDelay > clientConfigProperties.getReconnectMinDelay()
```

Comment:
this is very hard to read and understand, can we *untangle* this and use multiple expressions instead of a single one?

Reply:
Ok, I've split it up.